src/drivers/magnetometer/qmc5883/qmc5883.cpp

```diff
@@ -0,0 +1,1179 @@
+
+/* Control Register 1 */
+#define QMC5883_MODE_REG_STANDBY 		(0 << 0)
+#define QMC5883_MODE_REG_CONTINOUS_MODE         (1 << 0)
+#define QMC5883_OUTPUT_DATA_RATE_10 		(0 << 2) /* Hz */
+#define QMC5883_OUTPUT_DATA_RATE_50 		(1 << 2)
+#define QMC5883_OUTPUT_DATA_RATE_100 		(2 << 2)
+#define QMC5883_OUTPUT_DATA_RATE_200 		(3 << 2)
+#define QMC5883_OUTPUT_RANGE_2G 		(0 << 4)  /* +/- 2 gauss */
+#define QMC5883_OUTPUT_RANGE_8G 		(1 << 4)  /* +/- 8 gauss */
+#define QMC5883_OVERSAMPLE_512 			(0 << 6)  /* controls digital filter bw - larger OSR -> smaller bw */
+#define QMC5883_OVERSAMPLE_256 			(1 << 6)
+#define QMC5883_OVERSAMPLE_128 			(2 << 6)
+#define QMC5883_OVERSAMPLE_64 			(3 << 6)
+
+/* Control Register 2 */
+#define QMC5883_INT_ENB 			(1 << 0)
+#define QMC5883_ROL_PNT 			(1 << 6)
+#define QMC5883_SOFT_RESET 			(1 << 7)
+
+/* Set Register */
+#define QMC5883_SET_DEFAULT 			(1 << 0)
+
+#define QMC5883_TEMP_OFFSET 50 /* deg celsius */
+
+enum QMC5883_BUS {
+	QMC5883_BUS_ALL = 0,
+	QMC5883_BUS_I2C_INTERNAL,
+	QMC5883_BUS_I2C_EXTERNAL,
+	QMC5883_BUS_SPI
+};
+
+#ifndef CONFIG_SCHED_WORKQUEUE
+# error This requires CONFIG_SCHED_WORKQUEUE.
+#endif
+
+class QMC5883 : public device::CDev
+{
+public:
+	QMC5883(device::Device *interface, const char *path, enum Rotation rotation);
+	virtual ~QMC5883();
+
+	virtual int		init();
+
+	virtual ssize_t		read(struct file *filp, char *buffer, size_t buflen);
+	virtual int		ioctl(struct file *filp, int cmd, unsigned long arg);
+
+	/**
+	 * Stop the automatic measurement state machine.
+	 */
+	void			stop();
+
+	/**
+	 * Diagnostics - print some basic information about the driver.
+	 */
+	void			print_info();
+
+protected:
+	Device			*_interface;
+
+private:
+	work_s			_work{};
+	unsigned		_measure_ticks;
+
+	ringbuffer::RingBuffer	*_reports;
+	struct mag_calibration_s	_scale;
+	float 			_range_scale;
+	float 			_range_ga;
+	bool			_collect_phase;
+	int			_class_instance;
+	int			_orb_class_instance;
+
+	orb_advert_t		_mag_topic;
+
+	perf_counter_t		_sample_perf;
+	perf_counter_t		_comms_errors;
+	perf_counter_t		_range_errors;
+	perf_counter_t		_conf_errors;
+
+	/* status reporting */
+	bool			_sensor_ok;		/**< sensor was found and reports ok */
+
+	enum Rotation		_rotation;
+
+	struct mag_report	_last_report {};         /**< used for info() */
+
+	uint8_t			_range_bits;
+	uint8_t			_conf_reg;
+	uint8_t			_temperature_counter;
+	uint8_t			_temperature_error_count;
+
+	/**
+	 * Initialise the automatic measurement state machine and start it.
+	 *
+	 * @note This function is called at open and error time.  It might make sense
+	 *       to make it more aggressive about resetting the bus in case of errors.
+	 */
+	void			start();
+
+	/**
+	 * Reset the device
+	 */
+	int			reset();
+
+	/**
+	 * check the sensor configuration.
+	 *
+	 * checks that the config of the sensor is correctly set, to
+	 * cope with communication errors causing the configuration to
+	 * change
+	 */
+	void 			check_conf(void);
+
```

Comment:
Removed the calibration function - QMC5883 doesn't have a self-test function with an induced magnetic field like the HMC5983 has.